pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/AbstractMetadataStore.java

```diff
@@ -409,14 +409,19 @@ public void registerSessionListener(Consumer<SessionEvent> listener) {
 
     protected void receivedSessionEvent(SessionEvent event) {
         isConnected = event.isConnected();
-
-        sessionListeners.forEach(l -> {
-            try {
-                l.accept(event);
-            } catch (Throwable t) {
-                log.warn("Error in processing session event", t);
-            }
-        });
+        try {
+            executor.execute(() -> {
```

Comment:
One concern that I have with this patch is that usually in ZooKeeper client applications you leverage the feature that all the events are dispatched in strict order.
Here we are moving the execution of the listeners to another thread, and it actually works as the executor is still single threaded.
But on the other hand.. aren't we moving the problem to the `metadata-store` ?

maybe the problem is to be fixed in `AutoRecoveryMain.... onSessionExpired()` that is performing a blocking operation (depending on a sync ZooKeeper calll) on the main zookeeper thread.

Reply:
Good point. However, in AbstractMetadataStore, the other notifications are delivered using a separate thread pool: https://github.com/apache/pulsar/blob/d8c39bcb5becd15944978755c0b02fc52182fa69/pulsar-metadata/src/main/java/org/apache/pulsar/metadata/impl/AbstractMetadataStore.java#L273-L289
This PR makes AbstractMetadataStore more consistent by also delivering session events using this same thread pool.

Reply:
I noticed that call as well ;-) 

So overall I agree with this patch, because it makes it consistent the way we dispatch the events.

But I am not 100% sure that it will fix the original problem